code/modules/holoparasite/holoparasite_builder.dm

```diff
@@ -0,0 +1,572 @@
+				desc = "Amount of damage the $theme can negate, rather than transferring the entirety of it to the summoner.",
+				level = saved_stats.defense
+			),
+			list(
+				name = "Speed",
+				desc = "How fast the $theme can move and attack targets.",
+				level = saved_stats.speed
+			),
+			list(
+				name = "Potential",
+				desc = "Does nothing on its own, but it boosts the power of the $theme's abilities in various ways, although other stats can do so as well.",
+				level = saved_stats.potential
+			),
+			list(
+				name = "Range",
+				desc = "How far the $theme can travel from its summoner before being forcefully snapped back to the summoner's position.",
+				level = saved_stats.range
+			)
+		),
+		"validation" = list(
+			"color" = is_color_dark(accent_color, HOLOPARA_MAX_ACCENT_LIGHTNESS) ? "too dark" : "valid",
+			"name" = check_name_validity(),
+			"notes" = check_notes_validity()
+		)
+	)
+	if(length(notes))
+		.["notes"] = notes
+	if(saved_stats.ability)
+		.["selected_abilities"] += "[saved_stats.ability.type]"
+		if(saved_stats.ability.forced_weapon)
+			.["forced_weapon"] = "[saved_stats.ability.forced_weapon]"
+	for(var/datum/holoparasite_ability/ability as() in saved_stats.lesser_abilities)
+		.["selected_abilities"] += "[ability.type]"
+	.["selected_abilities"] += "[saved_stats.weapon.type]"
+
+
+/datum/holoparasite_builder/ui_static_data(mob/user)
+	. = list(
+		"themed_name" = theme.name,
+		"max_points" = max_points,
+		"max_level" = max_level,
+		"abilities" = list(
+			"major" = list(),
+			"lesser" = list(),
+			"weapons" = list()
+		),
+		"max_lengths" = list(
+			"name" = MAX_NAME_LEN,
+			"notes" = MAX_PAPER_LENGTH
+		)
+	)
+	for(var/ability_path in GLOB.holoparasite_abilities)
+		if(!is_valid_ability(ability_path))
+			continue
+		var/datum/holoparasite_ability/ability = GLOB.holoparasite_abilities[ability_path]
+		var/list/target_data
+		if(ispath(ability_path, /datum/holoparasite_ability/major))
+			target_data = .["abilities"]["major"]
+		else if(ispath(ability_path, /datum/holoparasite_ability/lesser))
+			target_data = .["abilities"]["lesser"]
+		else if(ispath(ability_path, /datum/holoparasite_ability/weapon))
+			target_data = .["abilities"]["weapons"]
+		else
+			continue
+		target_data += list(list(
+			name = ability.name,
+			desc = ability.desc,
+			cost = ability.cost,
+			icon = ability.ui_icon,
+			thresholds = ability.thresholds,
+			path = "[ability.type]",
+			hidden = ability.hidden
+		))
+
+/datum/holoparasite_builder/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	if(. || uses <= 0 || waiting)
+		return TRUE
+	switch(action)
+		if("set:name")
+			var/value = params["name"]
+			if(!istext(value))
+				return
+			holopara_name = trim(value, MAX_NAME_LEN)
+			. = TRUE
+		if("set:stat")
+			var/stat = params["stat"]
+			var/value = params["level"]
+			if(!istext(stat) || !length(stat) || !isnum_safe(value))
+				return
+			var/level = clamp(round(value), 1, max_level)
```

Comment:
Doesn't value have to be text2num'd at some point? Anything coming out of params is always going to be a text, no?
I'm not sure if something is missing here but I feel like this would have been caught up in testing if it was broken, so maybe byond is doing something other than the documented behaviours.

**DM reference: ** *Just remember that the parameter values are always stored as text, so you need to convert those to whatever data format you need using procedures such as text2num(), locate(), etc. *

Reply:
`params` is derived from decoded JSON sent back from tgui.